navigation/image_database.h

```diff
@@ -0,0 +1,173 @@
+#pragma once
+
+// BoB robotics includes
+#include "../common/pose.h"
+
+// Third-party includes
+#include "../third_party/path.h"
+#include "../third_party/units.h"
+
+// OpenCV includes
+#include <opencv2/opencv.hpp>
+
+// Standard C includes
+#include <cassert>
+#include <cstdio>
+
+// Standard C++ includes
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace BoBRobotics {
+namespace Navigation {
+using namespace units::length;
+using namespace units::angle;
+
+class ImageDatabase
+{
+public:
+    struct Entry {
+        Vector3<millimeter_t> position;
+        degree_t heading;
+        std::string path;
+
+        cv::Mat load() const
+        {
+            return cv::imread(path);
+        }
+
+        cv::Mat loadGreyscale() const
+        {
+            return cv::imread(path, cv::IMREAD_GRAYSCALE);;
+        }
+    };
+
+    template<class T>
+    ImageDatabase(T &&databaseName)
+      : m_Path(std::forward<T>(databaseName))
+    {
+        const auto metadataPath = m_Path / "metadata.csv";
+
+        // If we don't have metadata, it's an empty database
+        if (!metadataPath.exists()) {
+            // Make sure we have a directory to save into
+            filesystem::create_directory(m_Path);
+            return;
+        }
+
+        // Otherwise parse metadata file
+        std::ifstream file(metadataPath.str());
+        std::string line;
+        std::vector<std::string> fields;
+        std::getline(file, line); // skip first line
+        while (!file.eof()) {
```

Comment:
You can use ``std::getline`` to do the csv parsing bit too:
```c++
std::string lineString;
std::string fieldString;
while(std::getline(file, lineString)) {
    std::stringstream lineStream(lineString);
    fields.clear();
    while(std::getline(lineStream, fieldString ',')) {
        fields.push_back(fieldString);
    }
    
    //**TODO** turn fields into Entry
}
```

Reply:
Oh cool! That's good to know.